engines/gnap/gamesys.cpp

```diff
@@ -0,0 +1,1278 @@
+						if (animation->_framesCount > currFrameNum) {
+							while (animation->_framesCount > currFrameNum &&
+								animation->frames[currFrameNum]._duration <= duration) {
+								if (animation->frames[currFrameNum]._soundId != -1)
+									_soundIds.push_back((gfxItem->_sequenceId & 0xFFFF0000) | animation->frames[currFrameNum]._soundId);
+									duration -= animation->frames[currFrameNum]._duration;
+									++currFrameNum;
+							}
+							if (animation->_framesCount > currFrameNum)
+								gfxItem->_currFrame = animation->frames[currFrameNum++];
+							else
+								gfxItem->_currFrame = animation->frames[currFrameNum - 1];
+							if (gfxItem->_currFrame._spriteId != -1 && (seqItem->_x != 0 || seqItem->_y != 0))
+								gfxItem->_currFrame._rect.translate(seqItem->_x, seqItem->_y);
+							// Update sprite scaling
+							if ((seqItem->_flags & 1) && gfxItem->_currFrame._rect.bottom >= _backgroundImageValue1 && gfxItem->_currFrame._rect.bottom <= _backgroundImageValue3) {
+								int v17 = _backgroundImageValue2 + (gfxItem->_currFrame._rect.bottom - _backgroundImageValue1) *
+									(_backgroundImageValue4 - _backgroundImageValue2) /
+									(_backgroundImageValue3 - _backgroundImageValue1);
+								gfxItem->_currFrame._rect.top = gfxItem->_currFrame._rect.bottom - v17 * (gfxItem->_currFrame._rect.bottom - gfxItem->_currFrame._rect.top) / 1000;
+								gfxItem->_currFrame._rect.right = v17 * (gfxItem->_currFrame._rect.right - gfxItem->_currFrame._rect.left) / 1000 + gfxItem->_currFrame._rect.left;
+								gfxItem->_currFrame._isScaled = true;
+							}
+							gfxItem->_currFrame._duration -= duration;
+							if (gfxItem->_currFrame._soundId != -1)
+								_soundIds.push_back((gfxItem->_sequenceId & 0xFFFF0000) | gfxItem->_currFrame._soundId);
+							gfxItem->_currFrameNum = currFrameNum;
+							gfxItem->_updFlag = true;
+						} else if (v20 && gfxItem->_prevFrame._spriteId == -1) {
+							--_gfxItemsCount;
+							if (_gfxItemsCount != i)
+								memcpy(&_gfxItems[i], &_gfxItems[i + 1], sizeof(GfxItem) * (_gfxItemsCount - i));
+							--i;
+						} else {
+							gfxItem->_updFlag = false;
+						}
+					} else {
+						gfxItem->_prevFrame._duration -= duration;
+						gfxItem->_updFlag = false;
+					}
+				} else {
+					gfxItem->_delayTicks -= clockDelta;
+					gfxItem->_updFlag = false;
+				}
+			}
+		} else {
+			--_gfxItemsCount;
+			if (_gfxItemsCount != i)
+				memcpy(&_gfxItems[i], &_gfxItems[i + 1], sizeof(GfxItem) * (_gfxItemsCount - i));
+			--i;
+		}
+	}
+
+	if (_newSpriteDrawItemsCount > 0) {
+		debugC(kDebugBasic, "_newSpriteDrawItemsCount: %d", _newSpriteDrawItemsCount);
+		for (int k = 0; k < _newSpriteDrawItemsCount; ++k) {
+			if (_gfxItemsCount < 50) {
+				int insertIndex;
+				seqLocateGfx(-1, _newSpriteDrawItems[k]._id, &insertIndex);
+				if (_gfxItemsCount != insertIndex)
+					memcpy(&_gfxItems[insertIndex + 1], &_gfxItems[insertIndex], sizeof(GfxItem) * (_gfxItemsCount - insertIndex));
```

Comment:
I didn't get any Valgrind warning about overlapping memory areas for memcpy() here, but this looks like another case where we should use memmove() instead.